src/decode.rs

```diff
@@ -4542,40 +4553,28 @@ unsafe fn decode_b(
             case_set(bh4, &mut t.l, by4 as isize, set_ctx);
             case_set(bw4, &mut *t.a, bx4 as isize, set_ctx);
 
-            if (*f.frame_hdr).frame_type as libc::c_uint
-                & 1 as libc::c_int as libc::c_uint != 0
-            {
-                let r: *mut refmvs_block = &mut *(*(t.rt.r)
-                    .as_mut_ptr()
-                    .offset(
-                        ((t.by & 31 as libc::c_int) + 5 as libc::c_int + bh4
-                            - 1 as libc::c_int) as isize,
-                    ))
-                    .offset(t.bx as isize) as *mut refmvs_block;
-                let mut x: libc::c_int = 0 as libc::c_int;
-                while x < bw4 {
-                    (*r.offset(x as isize))
-                        .ref_0
-                        .ref_0[0 as libc::c_int as usize] = 0 as libc::c_int as int8_t;
-                    (*r.offset(x as isize)).bs = bs as uint8_t;
-                    x += 1;
+            if is_inter_or_switch(&*f.frame_hdr) {
+                let r: *mut refmvs_block = t
+                    .rt
+                    .r[((t.by & 31) + 5 + bh4 - 1) as usize]
+                    .offset(t.bx as isize);
+                for x in 0..bw4 {
+                    let block = &mut *r.offset(x as isize);
+                    block.ref_0.ref_0[0] = 0;
+                    block.bs = bs as uint8_t;
                 }
-                let mut rr: *const *mut refmvs_block = &mut *(t.rt.r)
-                    .as_mut_ptr()
-                    .offset(((t.by & 31 as libc::c_int) + 5 as libc::c_int) as isize)
-                    as *mut *mut refmvs_block;
-                let mut y: libc::c_int = 0 as libc::c_int;
-                while y < bh4 - 1 as libc::c_int {
-                    (*(*rr.offset(y as isize))
-                        .offset((t.bx + bw4 - 1 as libc::c_int) as isize))
-                        .ref_0
-                        .ref_0[0 as libc::c_int as usize] = 0 as libc::c_int as int8_t;
-                    (*(*rr.offset(y as isize))
-                        .offset((t.bx + bw4 - 1 as libc::c_int) as isize))
-                        .bs = bs as uint8_t;
-                    y += 1;
+
+                let mut rr: *const *mut refmvs_block = &t.rt.r[((t.by & 31) + 5) as usize];
+                for y in 0..bh4 - 1 {
+                    let block = &mut *rr
+                        .offset(y as isize)
+                        .read()
```

Comment:
This used to be a `*` deref, not a `ptr::read`.  I believe they're slightly different.

Reply:
Actually looks like there was a better way to express this that allows us to remove those extra raw pointer operations entirely.

Reply:
`ptr::read` is subtly different.  I looked into it and it *may* be the same for the specific type used here, but I think it's safest to stick with the `*` deref in general for these cases so we don't get bitten by one of those subtle differences.